airflow/providers/google/cloud/hooks/compute_ssh.py

```diff
@@ -0,0 +1,304 @@
+        if not self.instance_name or not self.zone or not self.project_id:
+            raise AirflowException(
+                f"Required parameters are missing: {missing_fields}. These parameters be passed either as "
+                "keyword parameter or as extra field in Airfow connection definition. Both are not set!"
+            )
+
+        self.log.info(
+            "Connecting to instance: instance_name=%s, user=%s, zone=%s, "
+            "use_internal_ip=%s, use_iap_tunnel=%s, use_os_login=%s",
+            self.instance_name,
+            self.user,
+            self.zone,
+            self.use_internal_ip,
+            self.use_iap_tunnel,
+            self.use_oslogin,
+        )
+        if not self.hostname:
+            hostname = self._compute_hook.get_instance_address(
+                zone=self.zone,
+                resource_id=self.instance_name,
+                project_id=self.project_id,
+                use_internal_ip=self.use_internal_ip or self.use_iap_tunnel,
+            )
+        else:
+            hostname = self.hostname
+
+        privkey, pubkey = self._generate_ssh_key(self.user)
+        if self.use_oslogin:
+            user = self._authorize_os_login(pubkey)
+        else:
+            user = self.user
+            self._authorize_compute_engine_instance_metadata(pubkey)
+
+        proxy_command = None
+        if self.use_iap_tunnel:
+            proxy_command_args = [
+                'gcloud',
+                'compute',
+                'start-iap-tunnel',
+                str(self.instance_name),
+                '22',
+                '--listen-on-stdin',
+                f'--project={self.project_id}',
+                f'--zone={self.zone}',
+                '--verbosity=warning',
+            ]
+            proxy_command = " ".join(shlex.quote(arg) for arg in proxy_command_args)
+
+        sshclient = self._connect_to_instance(user, hostname, privkey, proxy_command)
+        return sshclient
+
+    def _connect_to_instance(self, user, hostname, pkey, proxy_command) -> paramiko.SSHClient:
+        self.log.info("Opening remote connection to host: username=%s, hostname=%s", user, hostname)
+        for time_to_wait in exponential_sleep_generator(initial=1, maximum=10):
+            try:
+                client = _GCloudAuthorizedSSHClient(self._compute_hook)
+                # Default is RejectPolicy
+                # No known host checking since we are not storing privatekey
+                client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+                client.connect(
+                    hostname=hostname,
+                    username=user,
+                    pkey=pkey,
+                    sock=paramiko.ProxyCommand(proxy_command) if proxy_command else None,
+                    look_for_keys=False,
+                )
+                return client
+            except paramiko.SSHException:
+                if time_to_wait == 10:
```

Comment:
Can you explain it or leave a comment?

Reply:
exponential_sleep_generator is an infinite generator, so we need to manually check the end condition.
```
>>> pprint(list(zip(exponential_sleep_generator(1, 10), range(1, 30))))
[(0.3548877463516451, 1),
 (3.2179802309727807, 2),
 (6.267463665198731, 3),
 (10, 4),
 (10, 5),
 (10, 6),
 (10, 7),
 (10, 8),
 (10, 9),
 (10, 10),
 (10, 11),
 (10, 12),
 (10, 13),
 (10, 14),
 (10, 15),
 (10, 16),
 (10, 17),
 (10, 18),
 (10, 19),
 (10, 20),
 (10, 21),
 (10, 22),
 (10, 23),
 (10, 24),
 (10, 25),
 (10, 26),
 (10, 27),
 (10, 28),
 (10, 29)]
```

Reply:
Got it, should we create `max_wait_time` variable to keep in sync the `if` clause and generator params?